File: sdk/python/chronoguard_sdk/agents.py
"""Agent management API for ChronoGuard SDK.

This module provides methods for managing agents through the ChronoGuard API.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from chronoguard_sdk.models import Agent, AgentListResponse, CreateAgentRequest, UpdateAgentRequest

if TYPE_CHECKING:
    from httpx import AsyncClient, Client


class AgentsAPI:
    """Agent management API interface.

    Provides methods for creating, reading, updating, and listing agents.
    """

    def __init__(self, client: AsyncClient, base_url: str, tenant_id: str | None = None) -> None:
        """Initialize agents API.

        Args:
            client: HTTP client instance
            base_url: Base URL for API endpoints
            tenant_id: Optional default tenant ID
        """
        self._client = client
        self._base_url = base_url
        self._tenant_id = tenant_id

    async def list(
        self,
        page: int = 1,
        page_size: int = 50,
        status_filter: str | None = None,
    ) -> AgentListResponse:
        """List all agents for the tenant.

        Args:
            page: Page number (default: 1)
            page_size: Items per page (default: 50, max: 1000)
            status_filter: Optional status filter (e.g., "active", "inactive")

        Returns:
            Paginated list of agents

        Raises:
            ValidationError: If pagination parameters are invalid
            APIError: If the API request fails
        """
        params: dict[str, Any] = {"page": page, "page_size": page_size}
        if status_filter:
            params["status_filter"] = status_filter

        response = await self._client.get(f"{self._base_url}/api/v1/agents/", params=params)
        response.raise_for_status()
        return AgentListResponse(**response.json())

    async def get(self, agent_id: str) -> Agent:
        """Retrieve an agent by ID.

        Args:
            agent_id: Agent identifier

        Returns:
            Agent details

        Raises:
            NotFoundError: If agent not found
            APIError: If the API request fails
        """
        response = await self._client.get(f"{self._base_url}/api/v1/agents/{agent_id}")
        response.raise_for_status()
        return Agent(**response.json())

    async def create(
        self,
        name: str,
        certificate_pem: str,
        metadata: dict[str, Any] | None = None,
    ) -> Agent:
        """Create a new agent.

        Args:
            name: Agent name (3-100 characters)
            certificate_pem: X.509 certificate in PEM format
            metadata: Optional metadata dictionary

        Returns:
            Created agent details

        Raises:
            ValidationError: If validation fails
            ConflictError: If duplicate agent exists
            APIError: If the API request fails
        """
        request = CreateAgentRequest(
            name=name,
            certificate_pem=certificate_pem,
            metadata=metadata or {},
        )

        response = await self._client.post(
            f"{self._base_url}/api/v1/agents/",
            json=request.model_dump(mode="json"),
        )
        response.raise_for_status()
        return Agent(**response.json())

    async def update(
        self,
        agent_id: str,
        name: str | None = None,
        certificate_pem: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Agent:
        """Update an existing agent.

        Args:
            agent_id: Agent identifier
            name: Optional new agent name (3-100 characters)
            certificate_pem: Optional new X.509 certificate in PEM format
            metadata: Optional new metadata dictionary

        Returns:
            Updated agent details

        Raises:
            NotFoundError: If agent not found
            ValidationError: If validation fails
            APIError: If the API request fails
        """
        request = UpdateAgentRequest(
            name=name,
            certificate_pem=certificate_pem,
            metadata=metadata,
        )

        # Only include fields that are set
        request_data = request.model_dump(mode="json", exclude_none=True)

        response = await self._client.put(
            f"{self._base_url}/api/v1/agents/{agent_id}",
            json=request_data,
        )
        response.raise_for_status()
        return Agent(**response.json())


class AgentsSyncAPI:
    """Synchronous agent management API interface.

    Provides synchronous methods for creating, reading, updating, and listing agents.
    """

    def __init__(self, client: Client, base_url: str, tenant_id: str | None = None) -> None:
        """Initialize agents API.

        Args:
            client: HTTP client instance
            base_url: Base URL for API endpoints
            tenant_id: Optional default tenant ID
        """
        self._client = client
        self._base_url = base_url
        self._tenant_id = tenant_id

    def list(
        self,
        page: int = 1,
        page_size: int = 50,
        status_filter: str | None = None,
    ) -> AgentListResponse:
        """List all agents for the tenant.

        Args:
            page: Page number (default: 1)
            page_size: Items per page (default: 50, max: 1000)
            status_filter: Optional status filter (e.g., "active", "inactive")

        Returns:
            Paginated list of agents

        Raises:
            ValidationError: If pagination parameters are invalid
            APIError: If the API request fails
        """
        params: dict[str, Any] = {"page": page, "page_size": page_size}
        if status_filter:
            params["status_filter"] = status_filter

        response = self._client.get(f"{self._base_url}/api/v1/agents/", params=params)
        response.raise_for_status()
        return AgentListResponse(**response.json())

    def get(self, agent_id: str) -> Agent:
        """Retrieve an agent by ID.

        Args:
            agent_id: Agent identifier

        Returns:
            Agent details

        Raises:
            NotFoundError: If agent not found
            APIError: If the API request fails
        """
        response = self._client.get(f"{self._base_url}/api/v1/agents/{agent_id}")
        response.raise_for_status()
        return Agent(**response.json())

    def create(
        self,
        name: str,
        certificate_pem: str,
        metadata: dict[str, Any] | None = None,
    ) -> Agent:
        """Create a new agent.

        Args:
            name: Agent name (3-100 characters)
            certificate_pem: X.509 certificate in PEM format
            metadata: Optional metadata dictionary

        Returns:
            Created agent details

        Raises:
            ValidationError: If validation fails
            ConflictError: If duplicate agent exists
            APIError: If the API request fails
        """
        request = CreateAgentRequest(
            name=name,
            certificate_pem=certificate_pem,
            metadata=metadata or {},
        )

        response = self._client.post(
            f"{self._base_url}/api/v1/agents/",
            json=request.model_dump(mode="json"),
        )
        response.raise_for_status()
        return Agent(**response.json())

    def update(
        self,
        agent_id: str,
        name: str | None = None,
        certificate_pem: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Agent:
        """Update an existing agent.

        Args:
            agent_id: Agent identifier
            name: Optional new agent name (3-100 characters)
            certificate_pem: Optional new X.509 certificate in PEM format
            metadata: Optional new metadata dictionary

        Returns:
            Updated agent details

        Raises:
            NotFoundError: If agent not found
            ValidationError: If validation fails
            APIError: If the API request fails
        """
        request = UpdateAgentRequest(
            name=name,
            certificate_pem=certificate_pem,
            metadata=metadata,
        )

        request_data = request.model_dump(mode="json", exclude_none=True)

        response = self._client.put(
            f"{self._base_url}/api/v1/agents/{agent_id}",
            json=request_data,
        )
        response.raise_for_status()
        return Agent(**response.json())

File: sdk/javascript/src/agents.ts
/**
 * ChronoGuard Agent Management API
 *
 * Provides methods for managing agents (CRUD operations).
 */

import { HttpClient } from './client';
import {
  Agent,
  AgentListResponse,
  CreateAgentRequest,
  UpdateAgentRequest,
  AgentListOptions,
  UUID
} from './types';

/**
 * Agent management API client
 */
export class AgentAPI {
  private readonly client: HttpClient;

  constructor(client: HttpClient) {
    this.client = client;
  }

  /**
   * Create a new agent
   *
   * @param request - Agent creation request
   * @returns Created agent
   *
   * @throws {ValidationError} If request validation fails
   * @throws {ConflictError} If agent name already exists
   * @throws {ChronoGuardError} For other API errors
   *
   * @example
   * ```typescript
   * const agent = await client.agents.create({
   *   name: 'qa-agent-prod-01',
   *   certificate_pem: '-----BEGIN CERTIFICATE-----\n...\n-----END CERTIFICATE-----',
   *   metadata: { environment: 'production' }
   * });
   * console.log(`Created agent: ${agent.agent_id}`);
   * ```
   */
  async create(request: CreateAgentRequest): Promise<Agent> {
    return this.client.post<Agent, CreateAgentRequest>('/api/v1/agents/', request);
  }

  /**
   * Get an agent by ID
   *
   * @param agentId - Agent identifier
   * @returns Agent details
   *
   * @throws {NotFoundError} If agent is not found
   * @throws {ChronoGuardError} For other API errors
   *
   * @example
   * ```typescript
   * const agent = await client.agents.get('550e8400-e29b-41d4-a716-446655440000');
   * console.log(`Agent status: ${agent.status}`);
   * ```
   */
  async get(agentId: UUID): Promise<Agent> {
    return this.client.get<Agent>(`/api/v1/agents/${agentId}`);
  }

  /**
   * List all agents for the current tenant
   *
   * @param options - Pagination and filtering options
   * @returns Paginated list of agents
   *
   * @throws {ValidationError} If pagination parameters are invalid
   * @throws {ChronoGuardError} For other API errors
   *
   * @example
   * ```typescript
   * const response = await client.agents.list({
   *   page: 1,
   *   page_size: 50,
   *   status_filter: 'active'
   * });
   * console.log(`Total agents: ${response.total_count}`);
   * response.agents.forEach(agent => console.log(agent.name));
   * ```
   */
  async list(options: AgentListOptions = {}): Promise<AgentListResponse> {
    const params: Record<string, string | number> = {};

    if (options.page !== undefined) {
      params.page = options.page;
    }
    if (options.page_size !== undefined) {
      params.page_size = options.page_size;
    }
    if (options.status_filter !== undefined) {
      params.status_filter = options.status_filter;
    }

    return this.client.get<AgentListResponse>('/api/v1/agents/', { params });
  }

  /**
   * Update an existing agent
   *
   * @param agentId - Agent identifier
   * @param request - Update request (all fields optional)
   * @returns Updated agent
   *
   * @throws {NotFoundError} If agent is not found
   * @throws {ValidationError} If request validation fails
   * @throws {ChronoGuardError} For other API errors
   *
   * @example
   * ```typescript
   * const updatedAgent = await client.agents.update(
   *   '550e8400-e29b-41d4-a716-446655440000',
   *   { name: 'qa-agent-prod-02' }
   * );
   * console.log(`Updated agent: ${updatedAgent.name}`);
   * ```
   */
  async update(agentId: UUID, request: UpdateAgentRequest): Promise<Agent> {
    return this.client.put<Agent, UpdateAgentRequest>(`/api/v1/agents/${agentId}`, request);
  }
}
